### DISCUSSION TITLE
[Solved] How to bring a Brother legacy printer driver to the modern age?

### DISCUSSION THREAD

**Discussion by Ceaus** [OpenPrinting/cups#855] [Q&A] [answered]
I have a Brother DCP9020CDW laser printer. CUPS doesn't support this printer out of the box. Brother only supplies 32-bit RPM packages for this printer. Those RPMs are no longer supported on my distro (openSUSE Leap) since very recently.
I've unpacked those RPMs and they contain two 32-bit files (the rest is shell, data and PPD):
```
/opt/brother/Printers/dcp9020cdw/lpd/brdcp9020cdwfilter
/opt/brother/Printers/dcp9020cdw/cupswrapper/brcupsconfpt1
```

Is there anything I can do to bring this CUPS driver to the 21st century? Do I need those two files? Can those files be substituted with something else?

PS: For the moment I've setup a VM with an ancient openSUSE as printer server. And that works ok.

**Comment by zdohnal** [comment]
Hi,

what exactly do you mean by "Those RPMs are no longer supported on my distro"? The RPMs are directly from Brother, so they have never been supported by any distro - Brother was the one supporting the package, since they built it for specific distros and didn't provide source code, so none of open source distros could add it into official repositories. Someone could rip the binaries and files from Brother's RPMs and release it in a repository which accepts closed source licensed projects (f.e. repository "RPM Fusion non-free"), but IMO the support would still mainly lie on Brother.

It would be great if you mentioned what is the exact problem - package missing from repository? package installation failure? printing did not work after you installed the package and installed printer with correct driver?

In case you can unpack rpms and they just didn't install, you can put the files into correct directories by yourself - IMO it should work unless there is a crash in those binaries when you try to print...

**Reply by Ceaus** [reply]
Hi, 

Thanks for extensive reply. That is appreciated. Obviously the RPM's are provided by Brother. They were supported (installable) by my distro (openSUSE) as Brother uses the standard Linux FHS, and all required dependencies were present. With the latest two major releases it got already a bit more challenging due to missing dependencies. I had to fall back to Packman c.s.. The latest big update from openSUSE Leap 15.4 (a few weeks ago) had stuff removed and now I'm out of luck. The printer is no longer present (I suspect a critical dependency was removed). Regardless of what's available (or not) in openSUSE, I have to drag in a gazillion 32-bits compatibility libraries. As this laptop is my daily driver I'm not looking forward to solve the next to impossible. As a reminder, the printer driver dates back to 2013.

Point taken: I can unpack the RPMs and see if that confines the problem area, and reliefs me of installing 32-bit dependency crap I don't need nor want. I never though of doing that. I'll try this. Thanks for the suggestion. 

Moving back to my original question: is it possible to create a CUPS PPD  without those two 32-bit files? Or is that an insurmountable task?

**Reply by zdohnal** [reply, marked as answer]
First I will answer directly to your question, but IMHO you wanted something else.

Yes, you can create a CUPS PPD without those two files (one looks like a filter, the other as a conf tool) IF the printer supports a PDL (page description language - language which the printer is able to convert into its mechanic printing process) which can come from Linux printing system, e.g. Postscript or PDF - OR you have/can write a filter which will convert the incoming files into the specific PDL the printer accepts.

It depends how old the printer is and how "progressive" the manufacturer was at the time when the model was made - e.g. my HP from 2010 is capable of accepting straight PDFs (note that if you send a PDF to print does not mean a PDF arrives on the printer - this depends on driver). You will have to use PPD keywords and options when writing the PPD as well to define printer capabilities - here you can take inspiration from the original PPD - and add ```*cupsFilter2``` for converting the incoming file into the correct PDL.

However, IMHO you wanted to ask how to make the printer work without those 32bit files:

1. if the printer supports a correct version of IPP, you can try this - https://docs.fedoraproject.org/en-US/quick-docs/cups-useful-tricks/#_how_to_find_out_whether_my_printer_is_capable_of_driverless_printing - then you can either use temporary queue or install the printer as "everywhere" model,
2. if the printer supports PDL which CUPS provides a generic driver for, try [Ghostscript printer application](https://github.com/OpenPrinting/ghostscript-printer-app), either from SNAP or via github, or use the generic driver from CUPS directly - you can get the available models by ```lpinfo -m``` or in CUPS web interface during printer installation.
3. Or you can try drivers for similar models - [here](https://openprinting.org/printers/manufacturer/Brother) are some Brother DCPs from series 9000, those PPDs (not Postscript-Brother probably, I guess this will lead you back to 32b binaries you have) might do the trick as well,
4. if IPP, generic driver or similar do not work for you and the printer supports some classic PDL (if it uses IPP for communication, you can find it out via ipptool what ```document-format-supported``` - this is IPP attribute which reports supported document formats - which is similar to PDL, but more standardized, because it originates from MIME types. Otherwise model specification might mention the PDL), you can try to rewrite the existing PPD (you mentioned the Brother package had their own PPDs) to use e.g. Postscript by changing ```*cupsFilters2``` keyword and adjusting any possible non-standard PPD keywords (such keywords might be expected by the Brother filter and it won't be applied otherwise) - the original PPD standard which keywords are usually supported by CUPS is [here](https://web.archive.org/web/20160818204232/http://partners.adobe.com/public/developer/en/ps/5003.PPD_Spec_v4.3.pdf). 

To be honest, if IPP, generic or similar driver do not work for you, you have to prepare for difficult time and lot of testing and learning, so you have to decide whether it is worth it for you.

**Reply by Ceaus** [reply]
Thank you so  much. You've been very helpful. I'll research the options you've offered and see what gives.

**Comment by jsmeix** [comment]
Regarding the page description languages
for a Brother DCP9020CDW I found
https://support.brother.com/g/b/spec.aspx?c=eu_ot&lang=en&prod=dcp9020cdw_eu
that reads (excerpt):
```
Printer
...
Emulations 	PCL6, BR-Script3 (PostScript 3)
```

Accordingly at least for normal printing
(i.e. without special printer model specific features)
the Brother DCP9020CDW should "just work"
both as generic PCL printer
(often best for fast monochrome printout)
and as generic PostScript printer
(usually best for color printout but possibly slow)
by using a generic PCL PPD and/or a generic PostScript PPD.

I myself do not have a Brother printer.
I have a simple PostScript+PCL color laser printer
(i.e. only normal printing without special features)
for which I have set up two print queues
one as generic PCL5e printer with the 'ljet4' Ghostscript driver
and another one as generic PostScript printer (without driver),
cf. "Printer drivers: Make the printer print" on
https://en.opensuse.org/Concepts_printing

**Reply by Ceaus** [reply]
Thanks for the PCL/PS tips. That might be the simplest way to keep the printer running. On the OpenSUSE reference, that specific page wasn't on my reading list. Thanks!

**Comment by Ceaus** [comment]
Update: The first thing I tried was driverless printing using the direct IPP connection. It was first time right! I've got all the printer capabilities available: duplex, resolution, paper trays, etc. What a discovery to find out I've been depending on a printer driver I never needed in the first place...
This gives me yeeears of additional printer quality time :-) I can spin down my ancient openSUSE 32-bit printer server VM now.
Thank you CUPS!!!
